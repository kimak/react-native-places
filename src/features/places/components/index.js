import React, { Component } from "react";
import {
  Text,
  View,
  StyleSheet,
  ScrollView,
  Image,
  StatusBar
} from "react-native";

import PlaceList from "./PlaceList";
import Header from "./Header";
import Footer from "./Footer";

const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#ecf0f1"
  },
  content: {
    flex: 0,
    alignItems: "center"
  }
});

const Places = props => (
  <View style={styles.root}>
    <Header onAdd={props.addItem} />
    <ScrollView contentContainerStyle={styles.content}>
      <PlaceList places={props.places} onSelectItem={props.onToggleItem} />
    </ScrollView>
    <Footer onSelectFilter={props.updateFilter} selected={props.filter} />
  </View>
);
export default Places;
