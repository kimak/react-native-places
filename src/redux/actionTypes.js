const FETCH_PLACES_REQUEST = "FETCH_PLACES_REQUEST";
const FETCH_PLACES_SUCCESS = "FETCH_PLACES_SUCCESS";
const FETCH_PLACES_ERROR = "FETCH_PLACES_ERROR";
const ADD_PLACE = "ADD_PLACE";
const TOGGLE_PLACE = "TOGGLE_PLACE";
const SET_FILTER = "SET_FILTER";

export default {
  FETCH_PLACES_REQUEST,
  FETCH_PLACES_SUCCESS,
  FETCH_PLACES_ERROR,
  ADD_PLACE,
  TOGGLE_PLACE,
  SET_FILTER
};
