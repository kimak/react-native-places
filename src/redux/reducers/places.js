import initialStates from "../initialStates";
import actionsTypes from "../actionTypes";

function places(state = initialStates.places, action) {
  switch (action && action.type) {
    case actionsTypes.FETCH_PLACES_REQUEST:
      return { ...state, isLoading: true };
    case actionsTypes.FETCH_PLACES_SUCCESS:
      return { ...state, list: action.payload, isLoading: false };
    case actionsTypes.FETCH_PLACES_ERROR:
      return { ...state, error: action.payload, isLoading: false };
    case actionsTypes.ADD_PLACE:
      const list = state.list.concat();
      list.unshift({
        label: action.label,
        visited: false,
        id: list.length
      });
      return { ...state, list };
    case actionsTypes.TOGGLE_PLACE:
      return {
        ...state,
        list: state.list.map(
          todo =>
            todo.id === action.id ? { ...todo, visited: !todo.visited } : todo
        )
      };
    default:
      return state;
  }
}

export default places;
