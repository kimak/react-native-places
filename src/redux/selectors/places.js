import { createSelector } from "reselect";

export const filterPlaces = createSelector(
  (filter, places) => filter,
  (filter, places) => places,
  (visibilityFilter, places) => {
    switch (visibilityFilter) {
      case "all":
        return places;
      case "visited":
        return places.filter(t => t.visited);
      case "new":
        return places.filter(t => !t.visited);
    }
  }
);

const getFilter = state => state.placesFilter;
const getPlaces = state => state.places.list;

export const getVisiblePlaces = createSelector(
  [getFilter, getPlaces],
  (visibilityFilter, places) => {
    switch (visibilityFilter) {
      case "all":
        return places;
      case "visited":
        return places.filter(t => t.visited);
      case "new":
        return places.filter(t => !t.visited);
    }
  }
);
