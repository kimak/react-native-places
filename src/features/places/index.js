import places from "./containers/graphql";
export default places;
