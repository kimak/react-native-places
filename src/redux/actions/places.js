import actionTypes from "../actionTypes";

/* action creator used by redux-thunk middleware */
export function fetchPlaces() {
  return dispatch => {
    dispatch({ type: actionTypes.FETCH_PLACES_REQUEST });

    fetch("https://api.myjson.com/bins/r4r67")
      .then(response => {
        return response.json();
      })
      .then(payload => {
        dispatch({ type: actionTypes.FETCH_PLACES_SUCCESS, payload });
      })
      .catch(payload => {
        dispatch({ type: actionTypes.FETCH_PLACES_ERROR, payload });
      });
  };
}

export function togglePlace(id) {
  return { type: actionTypes.TOGGLE_PLACE, id };
}

export function addPlace(label) {
  return { type: actionTypes.ADD_PLACE, label };
}

export function updateFilter(filter) {
  return { type: actionTypes.SET_FILTER, filter };
}
