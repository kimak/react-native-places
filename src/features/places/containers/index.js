import React from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import {
  fetchPlaces,
  togglePlace,
  addPlace,
  updateFilter
} from "../../../redux/actions/places";
import Component from "../components";
import withDidMount from "../../../shared-ui/hoc/withDidMount";
import withLoading from "../../../shared-ui/hoc/withLoading";
import withError from "../../../shared-ui/hoc/withError";
import { getVisiblePlaces } from "../../../redux/selectors/places";

function mapDispatchToProps(dispatch) {
  return {
    didMount() {
      dispatch(fetchPlaces());
    },
    onToggleItem: id => {
      dispatch(togglePlace(id));
    },
    addItem: label => {
      dispatch(addPlace(label));
    },
    updateFilter: value => {
      dispatch(updateFilter(value));
    }
  };
}

function mapStateToProps(state) {
  return {
    places: getVisiblePlaces(state),
    isLoading: state.places.isLoading,
    error: state.places.error,
    filter: state.placesFilter
  };
}

const PlacesContainer = compose(
  connect(mapStateToProps, mapDispatchToProps),
  withDidMount,
  withError,
  withLoading
)(Component);
export default PlacesContainer;
