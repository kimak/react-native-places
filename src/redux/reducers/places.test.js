import places from "./places";
import actionTypes from "../actionTypes";

describe("places reducer", () => {
  it("snapshot test must return initialState", () => {
    expect(places()).toMatchSnapshot();
  });

  it(`snapshot test on action ${actionTypes.ADD_PLACE}`, () => {
    expect(
      places(undefined, { type: actionTypes.ADD_PLACE, label: "Paris" })
    ).toMatchSnapshot();
  });

  it(`snapshot test on action ${actionTypes.TOGGLE_PLACE}`, () => {
    expect(
      places(
        { list: [{ id: 0, label: "Paris", visited: false }] },
        {
          type: actionTypes.TOGGLE_PLACE,
          index: 0
        }
      )
    ).toMatchSnapshot();
  });
});
