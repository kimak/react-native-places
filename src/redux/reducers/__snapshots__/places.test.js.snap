// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`places reducer snapshot test must return initialState 1`] = `
Object {
  "error": undefined,
  "isLoading": true,
  "list": Array [],
}
`;

exports[`places reducer snapshot test on action ADD_PLACE 1`] = `
Object {
  "error": undefined,
  "isLoading": true,
  "list": Array [
    Object {
      "id": 0,
      "label": "Paris",
      "visited": false,
    },
  ],
}
`;

exports[`places reducer snapshot test on action TOGGLE_PLACE 1`] = `
Object {
  "list": Array [
    Object {
      "id": 0,
      "label": "Paris",
      "visited": true,
    },
  ],
}
`;
